File: api_service/src/services/commsec/secret_key_svc.py
from time import time
from typing import Any, cast

from src.models.auth.bootstrap import PublicKeyLookupResult
from src.models.commsec.commsec import (
    LocalPrivateKeyRef,
    ServiceKeyOwner,
    ServicePublicKeyRecord,
)
from src.repo.mysql_client import MySQLClient
from src.repo.mysql_dao import ServicePublicKeysDAO


class SecretKeyService:
    """密钥服务：本地私钥引用 + 全局公钥目录查询。"""

    def __init__(
        self,
        local_public_key: ServicePublicKeyRecord,
        local_private_key: LocalPrivateKeyRef,
        catalog: list[ServicePublicKeyRecord] | None = None,
        mysql_client: MySQLClient | None = None,
    ):
        self._local_public_key = self._normalize_record(local_public_key)
        self._local_private_key = local_private_key
        self._mysql_client = mysql_client
        self._public_keys_dao = (
            ServicePublicKeysDAO(mysql_client) if mysql_client else None
        )
        self._catalog_by_key_id: dict[str, ServicePublicKeyRecord] = {
            self._local_public_key.key_id: self._local_public_key,
        }
        for item in catalog or []:
            normalized = self._normalize_record(item)
            self._catalog_by_key_id[normalized.key_id] = normalized

    # TODO: 后续根据实际需求，更新函数内部实现，支持从安全存储加载或定期刷新本地公钥信息。
    async def get_public_key(self) -> ServicePublicKeyRecord:
        return self._local_public_key

    # TODO: 后续根据实际需求，更新函数内部实现，支持从安全存储加载或定期刷新本地私钥引用信息。
    async def get_private_key_ref(self) -> LocalPrivateKeyRef:
        return self._local_private_key

    async def get_public_key_by_key_id(self, key_id: str) -> PublicKeyLookupResult:
        normalized_key_id = key_id.strip()
        if not normalized_key_id:
            return PublicKeyLookupResult(
                found=False,
                key=None,
                failure_reason="key id is required",
                checked_at=time(),
            )

        key = self._catalog_by_key_id.get(normalized_key_id)
        if key is None:
            key = await self._load_public_key_by_id(normalized_key_id)
            if key is not None:
                self._catalog_by_key_id[key.key_id] = key
        if key is None:
            return PublicKeyLookupResult(
                found=False,
                key=None,
                failure_reason="key id not found",
                checked_at=time(),
            )
        return PublicKeyLookupResult(
            found=True,
            key=key,
            failure_reason="",
            checked_at=time(),
        )

    async def get_public_key_by_entity_id(
        self, entity_id: str
    ) -> PublicKeyLookupResult:
        normalized_entity_id = entity_id.strip()
        if not normalized_entity_id:
            return PublicKeyLookupResult(
                found=False,
                key=None,
                failure_reason="entity id is required",
                checked_at=time(),
            )

        key = self._pick_catalog_key_by_entity_id(normalized_entity_id)
        if key is None:
            key = await self._load_public_key_by_entity_id(normalized_entity_id)
            if key is not None:
                self._catalog_by_key_id[key.key_id] = key

        if key is None:
            return PublicKeyLookupResult(
                found=False,
                key=None,
                failure_reason="entity id not found",
                checked_at=time(),
            )

        return PublicKeyLookupResult(
            found=True,
            key=key,
            failure_reason="",
            checked_at=time(),
        )

    async def get_public_keys_by_owner(
        self, owner: ServiceKeyOwner
    ) -> list[ServicePublicKeyRecord]:
        owner = owner.normalized()
        out: list[ServicePublicKeyRecord] = []
        for key in self._catalog_by_key_id.values():
            if owner.owner_type and owner.owner_type != key.owner.owner_type:
                continue
            if owner.entity_type and owner.entity_type != key.owner.entity_type:
                continue
            if (
                owner.effective_entity_id
                and owner.effective_entity_id != key.owner.effective_entity_id
            ):
                continue
            if (
                owner.effective_entity_name
                and owner.effective_entity_name != key.owner.effective_entity_name
            ):
                continue
            if owner.instance_id and owner.instance_id != key.owner.instance_id:
                continue
            if owner.instance_name and owner.instance_name != key.owner.instance_name:
                continue
            out.append(key)
        if out:
            return out

        db_items = await self._load_public_keys_by_owner(owner)
        for item in db_items:
            self._catalog_by_key_id[item.key_id] = item
        return db_items

    async def _load_public_key_by_id(
        self, key_id: str
    ) -> ServicePublicKeyRecord | None:
        if self._public_keys_dao is None:
            return None
        row = await self._public_keys_dao.find_by_id(key_id)
        return self._row_to_record(row) if row else None

    async def _load_public_key_by_entity_id(
        self, entity_id: str
    ) -> ServicePublicKeyRecord | None:
        if self._public_keys_dao is None:
            return None

        rows = await self._public_keys_dao.find_many(
            filters={"entity_id": entity_id},
            order_by=["-activated_at", "-expires_at"],
            limit=8,
        )
        records = [
            record for row in rows if (record := self._row_to_record(row)) is not None
        ]
        if not records:
            return None

        preferred = records[0]
        for item in records[1:]:
            if self._should_prefer_key(item, preferred):
                preferred = item
        return preferred

    async def _load_public_keys_by_owner(
        self, owner: ServiceKeyOwner
    ) -> list[ServicePublicKeyRecord]:
        if self._public_keys_dao is None:
            return []

        filters: dict[str, str] = {}
        if owner.owner_type:
            filters["owner_type"] = owner.owner_type
        if owner.entity_type:
            filters["entity_type"] = owner.entity_type
        if owner.effective_entity_id:
            filters["entity_id"] = owner.effective_entity_id
        if owner.effective_entity_name:
            filters["entity_name"] = owner.effective_entity_name
        if owner.instance_id:
            filters["instance_id"] = owner.instance_id
        if owner.instance_name:
            filters["instance_name"] = owner.instance_name

        rows = await self._public_keys_dao.find_many(filters=filters or None)

        return [
            record for row in rows if (record := self._row_to_record(row)) is not None
        ]

    def _row_to_record(self, row: dict | None) -> ServicePublicKeyRecord | None:
        if not row:
            return None
        created_at = row["created_at"].timestamp() if row.get("created_at") else 0.0
        activated_at = (
            row["activated_at"].timestamp() if row.get("activated_at") else 0.0
        )
        expires_at = row["expires_at"].timestamp() if row.get("expires_at") else 0.0
        revoked_at = row["revoked_at"].timestamp() if row.get("revoked_at") else 0.0

        entity_type = str(row.get("entity_type") or "")
        entity_id = str(row.get("entity_id") or "")
        entity_name = str(row.get("entity_name") or "")

        return ServicePublicKeyRecord(
            key_id=str(row["key_id"]),
            owner=ServiceKeyOwner(
                owner_type=cast(Any, str(row["owner_type"])),
                entity_type=entity_type,
                entity_id=entity_id,
                entity_name=entity_name,
                service_id=entity_id,
                service_name=entity_name,
                instance_id=str(row.get("instance_id") or ""),
                instance_name=str(row.get("instance_name") or ""),
            ).normalized(),
            key_exchange_algorithm=cast(Any, str(row["key_exchange_algorithm"])),
            signature_algorithm=cast(Any, str(row["signature_algorithm"])),
            public_key_pem=str(row["public_key_pem"]),
            fingerprint=str(row["fingerprint"]),
            status=cast(Any, str(row["status"])),
            created_at=created_at,
            activated_at=activated_at,
            expires_at=expires_at,
            revoked_at=revoked_at,
        )

    def _normalize_record(self, item: ServicePublicKeyRecord) -> ServicePublicKeyRecord:
        return ServicePublicKeyRecord(
            key_id=item.key_id,
            owner=item.owner.normalized(),
            key_exchange_algorithm=item.key_exchange_algorithm,
            signature_algorithm=item.signature_algorithm,
            public_key_pem=item.public_key_pem,
            fingerprint=item.fingerprint,
            status=item.status,
            created_at=item.created_at,
            activated_at=item.activated_at,
            expires_at=item.expires_at,
            revoked_at=item.revoked_at,
        )

    def _pick_catalog_key_by_entity_id(
        self, entity_id: str
    ) -> ServicePublicKeyRecord | None:
        preferred: ServicePublicKeyRecord | None = None
        for key in self._catalog_by_key_id.values():
            if key.owner.effective_entity_id != entity_id:
                continue
            if preferred is None or self._should_prefer_key(key, preferred):
                preferred = key
        return preferred

    @staticmethod
    def _should_prefer_key(
        candidate: ServicePublicKeyRecord,
        current: ServicePublicKeyRecord,
    ) -> bool:
        if candidate.status == "active" and current.status != "active":
            return True
        if candidate.status != "active" and current.status == "active":
            return False
        if candidate.activated_at > current.activated_at:
            return True
        if candidate.expires_at > current.expires_at:
            return True
        return False

File: gateway/src/services/commsec/secret_key_svc.go
package commsec

import (
	"context"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"encoding/pem"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	commsecif "gateway/src/interfaces/commsec"
	commsecmodel "gateway/src/models/commsec"
	modelsystem "gateway/src/models/system"
	"gateway/src/repo"
)

var _ commsecif.ISecretKeyService = (*SecretKeyService)(nil)

// SecretKeyService 维护网关本地私钥引用和公钥目录缓存。
type SecretKeyService struct {
	mu sync.RWMutex

	localPublic  commsecmodel.ServicePublicKeyRecord
	localPrivate commsecmodel.LocalPrivateKeyRef
	catalogByKey map[string]commsecmodel.ServicePublicKeyRecord
	mysql        *repo.MySQLClient
}

// NewSecretKeyServiceFromConfig 基于本地密钥配置构建密钥服务。
// 该构造器负责装载并校验本地私钥/公钥，同时保留全局公钥目录查询能力。
func NewSecretKeyServiceFromConfig(
	cfg *modelsystem.SecretKeyConfig,
	catalog []commsecmodel.ServicePublicKeyRecord,
	mysql *repo.MySQLClient,
) (*SecretKeyService, error) {
	normalized := cfg.Normalized("gateway")
	if !normalized.Enabled {
		return NewSecretKeyServiceWithMySQL(
			commsecmodel.ServicePublicKeyRecord{},
			commsecmodel.LocalPrivateKeyRef{},
			catalog,
			mysql,
		), nil
	}

	if normalized.ActiveKeyID == "" {
		return nil, &modelsystem.ErrKeyIDRequired
	}

	publicRef := normalized.PublicKeyRef
	if publicRef == "" {
		publicRef = normalized.ActiveKeyID + ".public.pem"
	}
	privateRef := normalized.PrivateKeyRef
	if privateRef == "" {
		privateRef = normalized.ActiveKeyID + ".private.pem"
	}

	publicPEM, err := loadPEMBytesFromRef(publicRef, normalized.SecretDir)
	if err != nil {
		return nil, err
	}
	privatePEM, err := loadPEMBytesFromRef(privateRef, normalized.SecretDir)
	if err != nil {
		return nil, err
	}

	if err = ensureSPKIPublicKeyPEM(publicPEM); err != nil {
		return nil, err
	}
	detectedSignature, err := detectSignatureAlgorithm(publicPEM)
	if err != nil {
		return nil, err
	}
	if normalized.SignatureAlgorithm != "" && normalized.SignatureAlgorithm != string(detectedSignature) {
		return nil, fmt.Errorf(
			"signature algorithm mismatch: configured=%s detected=%s",
			normalized.SignatureAlgorithm,
			detectedSignature,
		)
	}
	if err = ensurePKCS8PrivateKeyPEM(privatePEM); err != nil {
		return nil, err
	}
	if err = ensurePrivateKeyMatchesAlgorithm(privatePEM, detectedSignature); err != nil {
		return nil, err
	}

	owner := commsecmodel.ServiceKeyOwner{
		OwnerType:    commsecmodel.CommKeyOwnerType(normalized.OwnerType),
		EntityType:   normalized.EntityType,
		EntityID:     normalized.EntityID,
		EntityName:   normalized.EntityName,
		ServiceID:    normalized.ServiceID,
		ServiceName:  normalized.ServiceName,
		InstanceID:   normalized.InstanceID,
		InstanceName: normalized.InstanceName,
	}.Normalized()

	now := time.Now()
	localPublic := commsecmodel.ServicePublicKeyRecord{
		KeyID:                normalized.ActiveKeyID,
		Owner:                owner,
		KeyExchangeAlgorithm: commsecmodel.KeyExchangeAlgorithm(normalized.KeyExchangeAlgorithm),
		SignatureAlgorithm:   detectedSignature,
		PublicKeyPEM:         string(publicPEM),
		Fingerprint:          sha256Hex(publicPEM),
		Status:               commsecmodel.CommKeyActive,
		CreatedAt:            now,
		ActivatedAt:          now,
		ExpiresAt:            time.Time{},
		RevokedAt:            time.Time{},
	}
	localPrivate := commsecmodel.LocalPrivateKeyRef{
		KeyID:                normalized.ActiveKeyID,
		Owner:                owner,
		KeyExchangeAlgorithm: commsecmodel.KeyExchangeAlgorithm(normalized.KeyExchangeAlgorithm),
		SignatureAlgorithm:   detectedSignature,
		PrivateKeyRef:        string(privatePEM),
		LoadedAt:             now,
	}

	return NewSecretKeyServiceWithMySQL(localPublic, localPrivate, catalog, mysql), nil
}

func NewSecretKeyService(
	localPublic commsecmodel.ServicePublicKeyRecord,
	localPrivate commsecmodel.LocalPrivateKeyRef,
	catalog []commsecmodel.ServicePublicKeyRecord,
) *SecretKeyService {
	return NewSecretKeyServiceWithMySQL(localPublic, localPrivate, catalog, nil)
}

func NewSecretKeyServiceWithMySQL(
	localPublic commsecmodel.ServicePublicKeyRecord,
	localPrivate commsecmodel.LocalPrivateKeyRef,
	catalog []commsecmodel.ServicePublicKeyRecord,
	mysql *repo.MySQLClient,
) *SecretKeyService {
	m := make(map[string]commsecmodel.ServicePublicKeyRecord)
	for _, item := range catalog {
		item.Owner = item.Owner.Normalized()
		if item.KeyID != "" {
			m[item.KeyID] = item
		}
	}
	localPublic.Owner = localPublic.Owner.Normalized()
	if localPublic.KeyID != "" {
		m[localPublic.KeyID] = localPublic
	}

	return &SecretKeyService{
		localPublic:  localPublic,
		localPrivate: localPrivate,
		catalogByKey: m,
		mysql:        mysql,
	}
}

// GetPublicKey 返回启动时装载并校验通过的本地公钥快照。
func (s *SecretKeyService) GetPublicKey(ctx context.Context) (commsecmodel.ServicePublicKeyRecord, error) {
	_ = ctx
	s.mu.RLock()
	defer s.mu.RUnlock()
	if s.localPublic.KeyID == "" {
		return commsecmodel.ServicePublicKeyRecord{}, &modelsystem.ErrLocalPublicKeyNotConfigured
	}
	return s.localPublic, nil
}

// GetPrivateKeyRef 返回启动时装载并校验通过的本地私钥引用。
func (s *SecretKeyService) GetPrivateKeyRef(ctx context.Context) (commsecmodel.LocalPrivateKeyRef, error) {
	_ = ctx
	s.mu.RLock()
	defer s.mu.RUnlock()
	if s.localPrivate.KeyID == "" {
		return commsecmodel.LocalPrivateKeyRef{}, &modelsystem.ErrLocalPrivateKeyRefNotConfigured
	}
	return s.localPrivate, nil
}

func (s *SecretKeyService) GetPublicKeyByKeyID(
	ctx context.Context, keyID string,
) (commsecmodel.PublicKeyLookupResult, error) {
	if keyID == "" {
		return commsecmodel.PublicKeyLookupResult{}, &modelsystem.ErrKeyIDRequired
	}
	s.mu.RLock()
	key, ok := s.catalogByKey[keyID]
	s.mu.RUnlock()
	if !ok && s.mysql != nil {
		dbKey, dbErr := s.loadPublicKeyByIDFromDB(ctx, keyID)
		if dbErr == nil && dbKey != nil {
			key = *dbKey
			ok = true
			s.mu.Lock()
			s.catalogByKey[keyID] = key
			s.mu.Unlock()
		}
	}
	if !ok {
		return commsecmodel.PublicKeyLookupResult{Found: false, FailureReason: "key id not found"}, nil
	}
	return commsecmodel.PublicKeyLookupResult{Found: true, Key: key}, nil
}

func (s *SecretKeyService) GetPublicKeyByEntityID(
	ctx context.Context, entityID string,
) (commsecmodel.PublicKeyLookupResult, error) {
	if entityID == "" {
		return commsecmodel.PublicKeyLookupResult{}, &modelsystem.ErrEntityIDRequired
	}

	s.mu.RLock()
	key, ok := pickCatalogKeyByEntityID(s.catalogByKey, entityID)
	s.mu.RUnlock()

	if !ok && s.mysql != nil {
		dbKey, dbErr := s.loadActivePublicKeyByEntityIDFromDB(ctx, entityID)
		if dbErr != nil {
			return commsecmodel.PublicKeyLookupResult{}, dbErr
		}
		if dbKey != nil {
			key = *dbKey
			ok = true
			s.mu.Lock()
			s.catalogByKey[key.KeyID] = key
			s.mu.Unlock()
		}
	}

	if !ok {
		return commsecmodel.PublicKeyLookupResult{Found: false, FailureReason: "entity id not found"}, nil
	}
	return commsecmodel.PublicKeyLookupResult{Found: true, Key: key}, nil
}

func (s *SecretKeyService) GetPublicKeysByOwner(
	ctx context.Context, owner commsecmodel.ServiceKeyOwner,
) ([]commsecmodel.ServicePublicKeyRecord, error) {
	owner = owner.Normalized()
	s.mu.RLock()
	items := make([]commsecmodel.ServicePublicKeyRecord, 0)
	for _, item := range s.catalogByKey {
		if !matchOwner(owner, item.Owner) {
			continue
		}
		items = append(items, item)
	}
	s.mu.RUnlock()

	if len(items) == 0 && s.mysql != nil {
		dbItems, dbErr := s.loadPublicKeysByOwnerFromDB(ctx, owner)
		if dbErr == nil {
			s.mu.Lock()
			for _, item := range dbItems {
				items = append(items, item)
				s.catalogByKey[item.KeyID] = item
			}
			s.mu.Unlock()
		}
	}
	return items, nil
}

func pickCatalogKeyByEntityID(
	catalogByKey map[string]commsecmodel.ServicePublicKeyRecord,
	entityID string,
) (commsecmodel.ServicePublicKeyRecord, bool) {
	var chosen commsecmodel.ServicePublicKeyRecord
	found := false
	for _, key := range catalogByKey {
		if key.Owner.EffectiveEntityID() != entityID {
			continue
		}
		if !found {
			chosen = key
			found = true
			continue
		}
		if shouldPreferKey(key, chosen) {
			chosen = key
		}
	}
	return chosen, found
}

func shouldPreferKey(candidate, current commsecmodel.ServicePublicKeyRecord) bool {
	if candidate.Status == commsecmodel.CommKeyActive && current.Status != commsecmodel.CommKeyActive {
		return true
	}
	if candidate.Status != commsecmodel.CommKeyActive && current.Status == commsecmodel.CommKeyActive {
		return false
	}
	if candidate.ActivatedAt.After(current.ActivatedAt) {
		return true
	}
	if candidate.ExpiresAt.After(current.ExpiresAt) {
		return true
	}
	return false
}

func (s *SecretKeyService) loadActivePublicKeyByEntityIDFromDB(
	ctx context.Context, entityID string,
) (*commsecmodel.ServicePublicKeyRecord, error) {
	if s.mysql == nil {
		return nil, &modelsystem.ErrNilMySQLClient
	}
	var row struct {
		KeyID                string    `db:"key_id"`
		OwnerType            string    `db:"owner_type"`
		EntityType           string    `db:"entity_type"`
		EntityID             string    `db:"entity_id"`
		EntityName           string    `db:"entity_name"`
		InstanceID           string    `db:"instance_id"`
		InstanceName         string    `db:"instance_name"`
		KeyExchangeAlgorithm string    `db:"key_exchange_algorithm"`
		SignatureAlgorithm   string    `db:"signature_algorithm"`
		PublicKeyPEM         string    `db:"public_key_pem"`
		Fingerprint          string    `db:"fingerprint"`
		Status               string    `db:"status"`
		CreatedAt            time.Time `db:"created_at"`
		ActivatedAt          time.Time `db:"activated_at"`
		ExpiresAt            time.Time `db:"expires_at"`
	}
	err := s.mysql.Get(ctx, &row, `
SELECT key_id, owner_type, entity_type, entity_id, entity_name, instance_id, instance_name,
       key_exchange_algorithm, signature_algorithm, public_key_pem, fingerprint,
       status, created_at, activated_at, expires_at
FROM auth_entity_public_keys
WHERE entity_id = ?
ORDER BY (status = 'active') DESC, activated_at DESC, expires_at DESC
LIMIT 1
`, entityID)
	if err != nil {
		if repo.IsNotFound(err) {
			return nil, nil
		}
		return nil, err
	}
	out := commsecmodel.ServicePublicKeyRecord{
		KeyID: row.KeyID,
		Owner: commsecmodel.ServiceKeyOwner{
			OwnerType:    commsecmodel.CommKeyOwnerType(row.OwnerType),
			EntityType:   row.EntityType,
			EntityID:     row.EntityID,
			EntityName:   row.EntityName,
			ServiceID:    row.EntityID,
			ServiceName:  row.EntityName,
			InstanceID:   row.InstanceID,
			InstanceName: row.InstanceName,
		}.Normalized(),
		KeyExchangeAlgorithm: commsecmodel.KeyExchangeAlgorithm(row.KeyExchangeAlgorithm),
		SignatureAlgorithm:   commsecmodel.SignatureAlgorithm(row.SignatureAlgorithm),
		PublicKeyPEM:         row.PublicKeyPEM,
		Fingerprint:          row.Fingerprint,
		Status:               commsecmodel.CommKeyStatus(row.Status),
		CreatedAt:            row.CreatedAt,
		ActivatedAt:          row.ActivatedAt,
		ExpiresAt:            row.ExpiresAt,
	}
	return &out, nil
}

func (s *SecretKeyService) loadPublicKeyByIDFromDB(
	ctx context.Context, keyID string,
) (*commsecmodel.ServicePublicKeyRecord, error) {
	if s.mysql == nil {
		return nil, &modelsystem.ErrNilMySQLClient
	}
	var row struct {
		KeyID                string    `db:"key_id"`
		OwnerType            string    `db:"owner_type"`
		EntityType           string    `db:"entity_type"`
		EntityID             string    `db:"entity_id"`
		EntityName           string    `db:"entity_name"`
		InstanceID           string    `db:"instance_id"`
		InstanceName         string    `db:"instance_name"`
		KeyExchangeAlgorithm string    `db:"key_exchange_algorithm"`
		SignatureAlgorithm   string    `db:"signature_algorithm"`
		PublicKeyPEM         string    `db:"public_key_pem"`
		Fingerprint          string    `db:"fingerprint"`
		Status               string    `db:"status"`
		CreatedAt            time.Time `db:"created_at"`
		ActivatedAt          time.Time `db:"activated_at"`
		ExpiresAt            time.Time `db:"expires_at"`
	}
	err := s.mysql.Get(ctx, &row, `
SELECT key_id, owner_type, entity_type, entity_id, entity_name, instance_id, instance_name,
       key_exchange_algorithm, signature_algorithm, public_key_pem, fingerprint,
       status, created_at, activated_at, expires_at
FROM auth_entity_public_keys
WHERE key_id = ?
LIMIT 1
`, keyID)
	if err != nil {
		if repo.IsNotFound(err) {
			return nil, nil
		}
		return nil, err
	}
	out := commsecmodel.ServicePublicKeyRecord{
		KeyID: row.KeyID,
		Owner: commsecmodel.ServiceKeyOwner{
			OwnerType:    commsecmodel.CommKeyOwnerType(row.OwnerType),
			EntityType:   row.EntityType,
			EntityID:     row.EntityID,
			EntityName:   row.EntityName,
			ServiceID:    row.EntityID,
			ServiceName:  row.EntityName,
			InstanceID:   row.InstanceID,
			InstanceName: row.InstanceName,
		}.Normalized(),
		KeyExchangeAlgorithm: commsecmodel.KeyExchangeAlgorithm(row.KeyExchangeAlgorithm),
		SignatureAlgorithm:   commsecmodel.SignatureAlgorithm(row.SignatureAlgorithm),
		PublicKeyPEM:         row.PublicKeyPEM,
		Fingerprint:          row.Fingerprint,
		Status:               commsecmodel.CommKeyStatus(row.Status),
		CreatedAt:            row.CreatedAt,
		ActivatedAt:          row.ActivatedAt,
		ExpiresAt:            row.ExpiresAt,
	}
	return &out, nil
}

func (s *SecretKeyService) loadPublicKeysByOwnerFromDB(
	ctx context.Context, owner commsecmodel.ServiceKeyOwner,
) ([]commsecmodel.ServicePublicKeyRecord, error) {
	owner = owner.Normalized()
	if s.mysql == nil {
		return nil, &modelsystem.ErrNilMySQLClient
	}
	query := `
SELECT key_id, owner_type, entity_type, entity_id, entity_name, instance_id, instance_name,
       key_exchange_algorithm, signature_algorithm, public_key_pem, fingerprint,
       status, created_at, activated_at, expires_at
FROM auth_entity_public_keys WHERE 1=1`
	args := make([]any, 0)
	if owner.OwnerType != "" {
		query += ` AND owner_type = ?`
		args = append(args, string(owner.OwnerType))
	}
	if owner.EntityType != "" {
		query += ` AND entity_type = ?`
		args = append(args, owner.EntityType)
	}
	if owner.EffectiveEntityID() != "" {
		query += ` AND entity_id = ?`
		args = append(args, owner.EffectiveEntityID())
	}
	if owner.EffectiveEntityName() != "" {
		query += ` AND entity_name = ?`
		args = append(args, owner.EffectiveEntityName())
	}
	if owner.InstanceID != "" {
		query += ` AND instance_id = ?`
		args = append(args, owner.InstanceID)
	}
	if owner.InstanceName != "" {
		query += ` AND instance_name = ?`
		args = append(args, owner.InstanceName)
	}

	type keyRow struct {
		KeyID                string    `db:"key_id"`
		OwnerType            string    `db:"owner_type"`
		EntityType           string    `db:"entity_type"`
		EntityID             string    `db:"entity_id"`
		EntityName           string    `db:"entity_name"`
		InstanceID           string    `db:"instance_id"`
		InstanceName         string    `db:"instance_name"`
		KeyExchangeAlgorithm string    `db:"key_exchange_algorithm"`
		SignatureAlgorithm   string    `db:"signature_algorithm"`
		PublicKeyPEM         string    `db:"public_key_pem"`
		Fingerprint          string    `db:"fingerprint"`
		Status               string    `db:"status"`
		CreatedAt            time.Time `db:"created_at"`
		ActivatedAt          time.Time `db:"activated_at"`
		ExpiresAt            time.Time `db:"expires_at"`
	}
	rows := make([]keyRow, 0)
	if err := s.mysql.Select(ctx, &rows, query, args...); err != nil {
		if repo.IsNotFound(err) {
			return nil, nil
		}
		return nil, err
	}
	out := make([]commsecmodel.ServicePublicKeyRecord, 0, len(rows))
	for _, r := range rows {
		if r.KeyID == "" {
			continue
		}
		out = append(out, commsecmodel.ServicePublicKeyRecord{
			KeyID: r.KeyID,
			Owner: commsecmodel.ServiceKeyOwner{
				OwnerType:    commsecmodel.CommKeyOwnerType(r.OwnerType),
				EntityType:   r.EntityType,
				EntityID:     r.EntityID,
				EntityName:   r.EntityName,
				ServiceID:    r.EntityID,
				ServiceName:  r.EntityName,
				InstanceID:   r.InstanceID,
				InstanceName: r.InstanceName,
			}.Normalized(),
			KeyExchangeAlgorithm: commsecmodel.KeyExchangeAlgorithm(r.KeyExchangeAlgorithm),
			SignatureAlgorithm:   commsecmodel.SignatureAlgorithm(r.SignatureAlgorithm),
			PublicKeyPEM:         r.PublicKeyPEM,
			Fingerprint:          r.Fingerprint,
			Status:               commsecmodel.CommKeyStatus(r.Status),
			CreatedAt:            r.CreatedAt,
			ActivatedAt:          r.ActivatedAt,
			ExpiresAt:            r.ExpiresAt,
		})
	}
	return out, nil
}

func matchOwner(expected commsecmodel.ServiceKeyOwner, actual commsecmodel.ServiceKeyOwner) bool {
	expected = expected.Normalized()
	actual = actual.Normalized()
	if expected.OwnerType != "" && expected.OwnerType != actual.OwnerType {
		return false
	}
	if expected.EntityType != "" && expected.EntityType != actual.EntityType {
		return false
	}
	if expected.EffectiveEntityID() != "" && expected.EffectiveEntityID() != actual.EffectiveEntityID() {
		return false
	}
	if expected.EffectiveEntityName() != "" && expected.EffectiveEntityName() != actual.EffectiveEntityName() {
		return false
	}
	if expected.InstanceID != "" && expected.InstanceID != actual.InstanceID {
		return false
	}
	if expected.InstanceName != "" && expected.InstanceName != actual.InstanceName {
		return false
	}
	return true
}

func loadPEMBytesFromRef(ref string, baseDir string) ([]byte, error) {
	raw := strings.TrimSpace(ref)
	if raw == "" {
		return nil, &modelsystem.ErrLocalPrivateKeyRefNotConfigured
	}
	if strings.Contains(raw, "-----BEGIN") {
		return []byte(raw), nil
	}
	if strings.HasPrefix(raw, "base64:") {
		encoded := strings.TrimSpace(strings.TrimPrefix(raw, "base64:"))
		if encoded == "" {
			return nil, &modelsystem.ErrInvalidPrivateKeyPEM
		}
		decoded, err := base64.StdEncoding.DecodeString(encoded)
		if err != nil {
			return nil, err
		}
		return decoded, nil
	}

	filePath := raw
	if after, ok := strings.CutPrefix(filePath, "file://"); ok {
		filePath = after
	}
	if !filepath.IsAbs(filePath) && baseDir != "" {
		filePath = filepath.Join(baseDir, filePath)
	}
	return os.ReadFile(filePath)
}

func ensureSPKIPublicKeyPEM(publicPEM []byte) error {
	block, _ := pem.Decode(publicPEM)
	if block == nil {
		return &modelsystem.ErrInvalidPublicKeyPEM
	}
	if _, err := x509.ParsePKIXPublicKey(block.Bytes); err != nil {
		return err
	}
	return nil
}

func ensurePKCS8PrivateKeyPEM(privatePEM []byte) error {
	block, _ := pem.Decode(privatePEM)
	if block == nil {
		return &modelsystem.ErrInvalidPrivateKeyPEM
	}
	if _, err := x509.ParsePKCS8PrivateKey(block.Bytes); err != nil {
		return err
	}
	return nil
}

func detectSignatureAlgorithm(publicPEM []byte) (commsecmodel.SignatureAlgorithm, error) {
	block, _ := pem.Decode(publicPEM)
	if block == nil {
		return "", &modelsystem.ErrInvalidPublicKeyPEM
	}
	parsed, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return "", err
	}

	switch pub := parsed.(type) {
	case ed25519.PublicKey:
		return commsecmodel.SignatureEd25519, nil
	case *ecdsa.PublicKey:
		if pub.Curve != nil && pub.Curve.Params().Name != "P-256" {
			return "", &modelsystem.ErrUnsupportedSignatureAlgorithm
		}
		return commsecmodel.SignatureECDSAP256SHA256, nil
	case *rsa.PublicKey:
		return commsecmodel.SignatureRSAPSSSHA256, nil
	default:
		return "", &modelsystem.ErrUnsupportedPublicKeyType
	}
}

func ensurePrivateKeyMatchesAlgorithm(
	privatePEM []byte,
	algorithm commsecmodel.SignatureAlgorithm,
) error {
	block, _ := pem.Decode(privatePEM)
	if block == nil {
		return &modelsystem.ErrInvalidPrivateKeyPEM
	}
	parsed, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return err
	}

	switch algorithm {
	case commsecmodel.SignatureEd25519:
		if _, ok := parsed.(ed25519.PrivateKey); !ok {
			return &modelsystem.ErrPrivateKeyNotEd25519
		}
		return nil
	case commsecmodel.SignatureECDSAP256SHA256:
		priv, ok := parsed.(*ecdsa.PrivateKey)
		if !ok {
			return &modelsystem.ErrPrivateKeyNotECDSA
		}
		if priv.Curve != nil && priv.Curve.Params().Name != "P-256" {
			return &modelsystem.ErrUnsupportedSignatureAlgorithm
		}
		return nil
	case commsecmodel.SignatureRSAPSSSHA256:
		if _, ok := parsed.(*rsa.PrivateKey); !ok {
			return &modelsystem.ErrPrivateKeyNotRSA
		}
		return nil
	default:
		return &modelsystem.ErrUnsupportedSignatureAlgorithm
	}
}

func sha256Hex(raw []byte) string {
	h := sha256.Sum256(raw)
	return hex.EncodeToString(h[:])
}
